# **Data Science Experience**
*Prof. Orlando Junior*

---

**Conteúdo**

1. Arranjos, vetores e matrizes
2. Listas
3. Tuplas
4. Sequências
5. Sets
6. Dicionários
7. Exercícios

---

## Arranjos, vetores e matrizes

A variável é capaz de armazenar apenas um valor de cada vez, mas existem situações em que há necessidade de armazenar uma grande quantidade de valores e para isso não iremos declarar e utilizar várias variáveis.

* Um **arranjo** (*array*) é um grupo/conjunto de elementos ou itens que respondem a um mesmo nome e que podem ser acessados segundo a posição que ocupam dentro do arranjo.
* Cada `elemento` de um arranjo está em determinada `posição` e armazena um `valor`, sendo que cada posição é associada a um `índice`.

![picture](https://ifrnead.github.io/rubynapratica/assets/images/estrutura-array.png)

Um arranjo não pode ter seu tamanho modificado durante a execução do algoritmo. Ele poderá ter várias dimensões:
* **Vetores** são arranjos unidimensionais (um índice);
* **Matrizes** são arranjos bidimensionais (dois ou mais índices);
* **Tensores** são arranjos multidimensionais.

**Vetores**

In [ ]:
# Vetor preenchido com 5 posições
vetor = [1, 2, 3, 4, 5]
print(vetor)

In [ ]:
# Tamanho
print("Tamanho: ", len(vetor))

In [ ]:
# Acessa cada elemento do vetor
for v in vetor:
  print(v, end=' ')

In [ ]:
# Acessa cada elemento do vetor pela posição
for idx in range(len(vetor)):
  print(vetor[idx], end=' ')

In [ ]:
# Modifica um valor pela posição
vetor[2] = 3000
print(vetor)

In [ ]:
# Criação de vetores com tipos de dado distintos
pessoa = [1, 'João', 2819.50, True]
print(pessoa)

In [ ]:
type(pessoa)

**Matrizes**

In [ ]:
# Matriz - 3 linhas x 4 colunas
pessoas = [[1, 'João', 2819.50, True],
           [2, 'Maria', 3459.76, False],
           [3, 'Carlos', 8294.77, False]]
print(pessoas)

In [ ]:
# Tamanho da matriz
print(len(pessoas))

In [ ]:
# Formato da matriz
print("Linhas : ", len(pessoas))
print("Colunas: ", len(pessoas[0]))

In [ ]:
# Modifica um valor da matriz
pessoas[0][0] = 1000
print(pessoas)

In [ ]:
# Imprime a matriz
atributos = ['Registro', 'Nome', 'Salário', 'CLT']

for i in range(len(pessoas)): # linhas
  for j in range(len(pessoas[i])): # colunas
    print("{0}: {1}".format(atributos[j], pessoas[i][j]))
  print()

In [ ]:
type(pessoas)

## Listas

**Documentação:** https://docs.python.org/3/library/stdtypes.html#lists

Uma **lista** é uma estrutura de dados que contém uma coleção ordenada de itens, ou seja, você pode armazenar uma sequência de itens em uma lista. Por exemplo, uma lista de compras.

In [ ]:
mercado = ['arroz', 'feijão', 'café']
print(mercado)

**Adicionar e remover elementos**

In [ ]:
# Adiciona no final lista
mercado.append('leite')
print(mercado)

In [ ]:
# Adiciona um elemento na posição 2 da lista
mercado.insert(2, 'queijo')
print(mercado)

In [ ]:
# Remove o elemento pelo nome, se existir
mercado.remove('arroz')
print(mercado)

In [ ]:
# Remove pela posição e devolve o elemento
mercado.pop(2)

In [ ]:
print(mercado)

In [ ]:
type(mercado)

**Iteradores e geradores**

In [ ]:
# Iterador como lista
faixa_valores = range(10)
print(faixa_valores)

In [ ]:
type(faixa_valores)

In [ ]:
list(faixa_valores)

In [ ]:
# Função geradora de números
def gera_numeros(n):
  i = 1
  while i <= n:
    yield i
    i = i + 1

In [ ]:
# Cria um gerador de números
gerador = gera_numeros(1000)
print(gerador)

In [ ]:
list(gerador)

In [ ]:
# Consome o gerador de números
for i in gera_numeros(1000):
  print(i, end=' ')

**Concateção e combinação**

In [ ]:
# Concatena usando '+'
lista1 = [1, 2, 3, 4, 5]
lista2 = [11, 12, 13, 14, 15]

In [ ]:
lista3 = lista1 + lista2
print(lista3)

In [ ]:
# Concatena com 'lista2' agrega na 'lista1' (mais rápido)
lista1.extend(lista2)
print(lista1)

**Ordenação**

In [ ]:
# Ordenação comum (sem chave)

In [ ]:
numeros = [3, 2, 4, 1, 5]

In [ ]:
numeros.sort()

In [ ]:
numeros

In [ ]:
# Ordenação por chave (por exemplo, tamanho do texto)
pessoas = ['João Pedro', 'Maria', 'Carlos', 'Ana']

In [ ]:
pessoas.sort(key=len, reverse=True)

In [ ]:
pessoas

**Fatiamento (*slicing*)**

In [ ]:
idades = [11, 13, 16, 12, 15, 17, 12, 14, 17]

In [ ]:
# Seleciona parte da lista
idades[1:3] # start:stop (exclusive)

In [ ]:
# Modifica parte da lista
idades[1:3] = [46, 37]

In [ ]:
print(idades)

In [ ]:
# Omite 'start'
idades[:5]      # Recupera os 5 primeiros elementos

In [ ]:
# Omite 'stop'
idades[3:]      # Recupera a partir da posição 3

In [ ]:
# Omite 'start' e 'stop'
idades[:]

In [ ]:
# Observa a lista do final (índices negativos)
print("Lista             :", idades)
print("Último elemento   :", idades[-1])
print("Penúltimo elemento:", idades[-2])

In [ ]:
# Fatiamento com índices negativos
print("Lista              :", idades)
print("------------------->", idades[-4:-1]) # Recua 4 elementos do fim e avança até o elemento 1
print("------------------->", idades[-5:-1]) # Recua 5 elementos do fim e avança até o elemento 1
print("------------------->", idades[-5:-2]) # Recua 5 elementos do fim e avança até o elemento 2
print("------------------->", idades[-6:-3]) # Recua 6 elementos do fim e avança até o elemento 3

In [ ]:
# Fatiamento alternado usando 'step'
cidades = ['Santo André', 'São Paulo', 'Lisboa',
           'Florianópolis', 'Gramado', 'Santiago',
           'Buenos Aires', 'Madrid']

In [ ]:
# Fatia do começo e seleciona a cada 2 itens
cidades[::2]

In [ ]:
# Fatia a partir do elemento 3
cidades[3::]

In [ ]:
# Fatia a partir do elemento 1 e seleciona a cada 3 itens
cidades[1::3]

***List comprehensions***: permitem que você componha uma nova lista de modo conciso, filtrando os elementos de uma coleção, transformando os elementos ao passar o filtro, com uma expressão concisa.

In [ ]:
salarios = [2790.0, 3250.0, 1260.0, 5200.0, 2440.0, 8670.0, 2010.0]
print(salarios)

In [ ]:
# Modo tradicional
# Seleciona os salários maiores que 2500.0
salarios_maiores2500 = []
for sal in salarios:
  if sal > 2500:
    salarios_maiores2500.append(sal)

In [ ]:
print(salarios_maiores2500)

In [ ]:
# Modo pythônico
# Seleciona os salários maiores que 2500.0
salarios_maiores2500 = [sal for sal in salarios if sal > 2500.0]
print(salarios_maiores2500)

In [ ]:
print(salarios)

In [ ]:
# Acrescenta 20% nos salários menores que 2500.0
novos_salarios = [sal * 1.2 if sal < 2500.0 else sal for sal in salarios]
print(novos_salarios)

In [ ]:
# Soma todos os novos salários menores que 2500.0
soma_salarios = sum([sal for sal in novos_salarios if sal < 2500.0])
print(soma_salarios)

**Expressões *lambda*:** são funções anônimas, pequenas, restritas e concisas.

In [ ]:
# Modo tradicional
# Eleva ao quadrado o valor recebido
def quadrado(n):
  return n ** 2

In [ ]:
quadrado(10)

In [ ]:
# Modo pythônico usando expressões lambda
# Elevar os valores da lista o quadrado
quadrado = lambda n : n**2

In [ ]:
quadrado(10)

## Tuplas

**Documentação:** https://docs.python.org/3/library/stdtypes.html#lists

Uma **tupla** é uma coleção de objetos que se mantêm ordenados e imutáveis.
* Tuplas são usadas para manter vários objetos juntos;
* Elas são semelhantes às listas, mas sem a extensa funcionalidade que a lista oferece;
* Assim como os literais, as tuplas são imutáveis , ou seja, você não pode modificá-las.
* Em geral, são usadas nos casos em que uma instrução ou função definida pelo usuário pode assumir com segurança que a coleção de valores não será alterada.

In [ ]:
# Cria uma tupla vazia
tupla = ()
print(tupla)

In [ ]:
# Tupla preenchida
valores = (1, 2, 3)
print(valores)

In [ ]:
# Tamanho
print(len(valores))

In [ ]:
# Tuplas aninhadas
valores = ((1, 2, 3), (4, 5, 6))
print(valores)

In [ ]:
# Converte uma sequência em uma tupla
lista1 = [100, 200, 300]
tuple(lista1)

In [ ]:
# Tupla a partir de texto
nome = tuple('João Pereira')
print(nome)

In [ ]:
# Acessa os itens da tupla por índices
print(nome[2])

In [ ]:
# Frequência de itens
numeros = (1, 1, 1, 2, 3, 4, 5)
print(numeros.count(1))

**Imutabilidade:** os objetos inseridos na tupla podem ser mutáveis, mas depois que a tupla é criada, não será possível modificar qualquer objeto armazenado em cada posição.

In [ ]:
valores = tuple(['João', 28, [84], True])
print(valores)

In [ ]:
valores[2] = 1.79 # erro

In [ ]:
valores[2].append(1.79)
print(valores)

**Desempacotamento:** as tuplas são uma maneira eficaz de retornar múltiplos valores a partir de funções.

In [ ]:
# Função soma_e_multiplica
def soma_e_multiplica(x, y):
  return (x + y), (x * y)

In [ ]:
resultado_soma, resultado_multiplicacao = soma_e_multiplica(3, 9)

In [ ]:
print("Soma      : ", resultado_soma)
print("Multiplica: ", resultado_multiplicacao)

**Desempacotamento sofisticado**

In [ ]:
zoo = ('cobra', 'elefante', 'pato', 'lagosta', 'caranguejo', 'aranha')

In [ ]:
reptil, mamifero, ave, *invertebrados = zoo

In [ ]:
print("Réptil       : ", reptil)
print("Mamífero     : ", mamifero)
print("Ave          : ", ave)
print("Invertebrados: ", invertebrados)

In [ ]:
# Forma alternativa
numeros = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

In [ ]:
a, b, c, *_ = numeros

In [ ]:
print(a)
print(b)
print(c)
print(_)

## Sequências

**Documentação:** https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range

Listas, tuplas e cadeias de caracteres (strings) são exemplos de **sequências**. Os principais recursos são os testes de associação (expressões `in` e `not in`) e as operações de indexação, que nos permitem buscar um item específico na sequência diretamente.

**enumerate:** devolve uma sequência de tuplas para que possamos manter o controle do índice do item durante as iterações.

In [ ]:
lista = ['O Poderoso Chefão', 'A Lista de Schindler', 'Um Sonho de Liberdade', 'O Rei Leão', 'O Auto da Compadecida']

In [ ]:
dicionario = {}

In [ ]:
for index, value in enumerate(lista):
  dicionario[index] = value

In [ ]:
dicionario

In [ ]:
# Formatação mais comum
dicionario = {}
for index, value in enumerate(lista):
  dicionario[value] = index

In [ ]:
dicionario

**sorted:** devolve uma lista ordenada.

In [ ]:
lista = [99, 0, 78, 2, 12, 0, 42, 56, 9, 0]

In [ ]:
sorted(lista)

**zip:** cria pares de elementos em uma série de sequências para criar uma lista de tuplas.

In [ ]:
paises = ['China', 'Índia', 'Estados Unidos', 'Indonésia', 'Paquistão'] # lista
populacao = (1411780000, 1380004385, 331449281, 273523615, 220892340)   # tupla

In [ ]:
paises_populacao = zip(paises, populacao)

In [ ]:
list(paises_populacao)

**reversed:** itera pelos elementos de uma sequência na ordem inversa.

In [ ]:
numeros = tuple(range(1, 21))
print(numeros)

In [ ]:
numeros_reversos = tuple(reversed(numeros))
print(numeros_reversos)

## *Sets*

**Documentação:** https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset

***Sets*** (conjuntos) são coleções não ordenadas de objetos simples. Eles são usados quando a existência de um objeto em uma coleção é mais importante do que a ordem ou quantas vezes ela ocorre.

* Usando conjuntos, você pode testar o pertencimento dos elementos, averiguar se um subconjunto pertence a um conjunto e também encontrar a intersecção entre dois conjuntos.
* Podemos pensar neles como dicionários, mas somente com as chaves, sem os valores.

In [ ]:
# Cria um set
brics = set(['Brasil', 'Índia', 'Rússia', 'China', 'África do Sul'])
print(brics)

In [ ]:
# Cria um set
mercosul = {'Brasil', 'Uruguai', 'Paraguai', 'Argentina', 'Venezuela'}
print(mercosul)

In [ ]:
# Verifica o tipo de dado
type(mercosul)

**Operações em conjuntos**

In [ ]:
# União
brics.union(mercosul)

In [ ]:
brics | mercosul

In [ ]:
# Intersecção
brics.intersection(mercosul)

In [ ]:
brics & mercosul

In [ ]:
# Diferença
brics.difference(mercosul)

In [ ]:
brics - mercosul

In [ ]:
# Contido (é subconjunto)
paises = brics & mercosul # Brasil
paises.issubset(brics)

## Dicionários

**Documentação:** https://docs.python.org/3/library/stdtypes.html#mapping-types-dict

Um **dicionário** é como uma agenda onde você pode encontrar o endereço ou detalhes de contato de uma pessoa sabendo apenas seu nome.
* Nos dicionários, associamos `chaves` com `valores`;
* A **chave** deve ser exclusiva e única para o dicionário, e você pode usar apenas objetos imutáveis para ela.



In [ ]:
# Cria um dicionário vazio
dicionario = dict()

In [ ]:
type(dicionario)

In [ ]:
# Cria um dicionário vazio no modo pythônico
dicionario = {}

In [ ]:
# Cria um dicionário com valores
dicionario = {'Nome': 'João', 'Idade': 29, 'Salário': 2700.00, 'Filhos': ['João', 'Maria', 'Marcos']}

In [ ]:
print(dicionario)

In [ ]:
# Acesso o valor pelo nome da chave
print(dicionario['Nome'])

**Operações**

In [ ]:
# Tamanho
len(dicionario)

In [ ]:
# Transforma em literal
str(dicionario)

In [ ]:
# Recupera o valor pelo nome da chave
dicionario.get('Nome')

In [ ]:
# Recupera a lista de itens
dicionario.items()

In [ ]:
# Recupera a lista de chaves
dicionario.keys()

In [ ]:
# Recupera a lista de valores
dicionario.values()

**Incluir e remover chaves**

In [ ]:
# Inclui uma nova chave com valor
dicionario['Peso'] = 75.6

In [ ]:
print(dicionario)

In [ ]:
# Remove uma chave pelo nome
del dicionario['Peso']

In [ ]:
print(dicionario)

In [ ]:
# Remove uma chave e devolve o valor associado
dicionario.pop('Idade')

In [ ]:
print(dicionario)

**Criando dicionários a partir de listas**

In [ ]:
lista1 = ['a', 'b', 'c', 'd']
lista2 = [10, 20, 30, 40]

In [ ]:
dados = {}

In [ ]:
# Usando iteradores
for key, value in zip(lista1, lista2):
  dados[key] = value

In [ ]:
print(dados)

In [ ]:
# Usando a função dict
dados_novos = dict(zip(lista1, lista2))

In [ ]:
print(dados_novos)

## Exercícios

**1 #** Crie um lista com os números pares entre 0 e 1000 e mostre essa lista.

In [1]:
list1 = [x * 2 for x in range(501)]
print(list1)

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268, 270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294, 296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320, 322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 394, 396, 398, 400, 402, 404, 406, 408, 410, 412, 414, 416, 418, 420,

**2 #** Elabore um algoritmo que apresente a saída abaixo como resultado da concatenação de três listas.

[10, 20, 30, 1, 2, 3, 10, 200, 300]


**3 #** Elabore um algoritmo que apresente a saída abaixo como resultado do fatiamento de uma lista de números ímpares entre 1 e 50.

[13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35]


**4 #** Elabore um algoritmo que apresente a saída abaixo como resultado de uma tupla criada a partir de nomes de animais domésticos.

('Cachorro', 'Gato', 'Calopsita', 'Hamster', 'Coelho', 'Tartaruga')


**5 #** Elabore um algoritmo que utilize a função `zip` para criar os pares de elementos apresentados na saída abaixo.

[('Petrobrás', 80000), ('Ambev', 120000), ('Nubank', 300000), ('Carrefour', 700000), ('JBS', 550000)]


**6 #** Elabore um algoritmo que utilize a função `enumerate` para criar o mapeamento numérico da lista de livros apresentada na saída em um dicionário.

{'Bíblia': 0, 'Pai Rico, Pai Pobre': 1, 'Python para Análise de Dados': 2, 'A Revolução dos Bichos': 3, 'Sentimento do Mundo': 4}


**7 #** Elabore um algoritmo que faça a intersecção de dois conjuntos sobre os ganhadores brasileiros no futebol sulamericano e apresente conforme a saída abaixo.

Brasileirão :  {'Cruzeiro', 'Palmeiras', 'São Paulo', 'Santos', 'Grêmio'}
Libertadores:  {'Palmeiras', 'São Paulo', 'Atlético Mineiro', 'Grêmio', 'Flamengo'}
Inserseção  :  {'Grêmio', 'Palmeiras', 'São Paulo'}


**8 #** Crie um dicionário com suas informações: nome, idade, empresa e time de futebol. Caso não torça para nenhum time, utilize a palavra `None`, conforme o exemplo apresentado na saída abaixo.

{'Nome': 'Orlando Junior', 'Idade': 17, 'Empresa': 'Mineradores', 'Time': None}


**9 #** Elabore um algoritmo para somar e exibir todos os valores do dicionário criado abaixo.

In [ ]:
mercado = {'arroz': 29.90, 'feijão': 8.90, 'açúcar': 3.49, 'sal': 2.49}

44.78

**10 #** Elabore um algoritmo para calcular o salário de uma pessoa dado o valor hora em um mês com 160 horas usando funções lambda.

8000